File: src/TopiaForge.ModManager.Core/LaunchProfile.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace TopiaForge.ModManager.Core
{
    /// <summary>An exact package selection whose manifest is captured independently of caller state.</summary>
    public sealed class ResolvedPackage
    {
        private readonly ModManifest snapshot;

        public ResolvedPackage(string id, string version, ModManifest manifest)
        {
            Identity = new PackageIdentity(id, version);
            if (manifest == null) throw new ArgumentNullException(nameof(manifest));
            if (!string.Equals(id, manifest.Id, StringComparison.OrdinalIgnoreCase) || version != manifest.Version)
                throw new ArgumentException("Selected identity must agree with its manifest.", nameof(manifest));
            snapshot = ModManifestJson.CopyForLaunch(manifest);
        }

        public PackageIdentity Identity { get; }
        public string Id => Identity.Id;
        public string Version => Identity.Version;
        public ModManifest Manifest => ModManifestJson.CopyForLaunch(snapshot);
        internal ModManifest Snapshot => snapshot;
    }

    public sealed class InstallFacts
    {
        public InstallFacts(string platform = "", string architecture = "", string contentTarget = "", string gameVersion = "")
        {
            Platform = platform ?? string.Empty;
            Architecture = architecture ?? string.Empty;
            ContentTarget = contentTarget ?? string.Empty;
            GameVersion = gameVersion ?? string.Empty;
        }

        public string Platform { get; }
        public string Architecture { get; }
        public string ContentTarget { get; }
        public string GameVersion { get; }
    }

    /// <summary>The exact enabled selection and selected disabled packages, without registry or filesystem access.</summary>
    public sealed class EffectiveProfile
    {
        public EffectiveProfile(string profileId, int revision, IReadOnlyList<ResolvedPackage> packages,
            InstallFacts? install = null, IReadOnlyList<ResolvedPackage>? disabledPackages = null)
        {
            ProfileId = LaunchContractValues.Token(profileId, nameof(profileId));
            Revision = LaunchContractValues.Revision(revision, nameof(revision));
            Packages = LaunchContractValues.Copy(packages);
            DisabledPackages = LaunchContractValues.Copy(disabledPackages);
            Install = install ?? new InstallFacts();
        }

        public string ProfileId { get; }
        public int Revision { get; }
        public IReadOnlyList<ResolvedPackage> Packages { get; }
        public IReadOnlyList<ResolvedPackage> DisabledPackages { get; }
        public InstallFacts Install { get; }
    }

    public static partial class ModManifestJson
    {
        internal static ModManifest CopyForLaunch(ModManifest manifest)
        {
            var copy = JsonUtil.Clone(manifest);
            NormalizeCollections(copy);
            NormalizeContributions(copy);
            return copy;
        }
    }
}

File: packages/launcher_domain/lib/src/launch_resolution/launch_profile.dart
part of '../launch_resolution.dart';

/// A copied identity; no manifest or caller-owned collection can mutate it.
final class PackageIdentity {
  PackageIdentity({required String id, required String version})
    : id = _packageId(id),
      version = _version(version);
  factory PackageIdentity.fromJson(Object? value) {
    final json = _object(value, {'id', 'version'}, {'id', 'version'});
    return PackageIdentity(
      id: _packageId(json['id']),
      version: _version(json['version']),
    );
  }
  final String id;
  final String version;
  Map<String, Object?> toJson() => {'id': id, 'version': version};
}

/// Captures the installed manifest at construction, returning defensive copies.
final class ResolvedPackage extends PackageIdentity {
  ResolvedPackage({
    required super.id,
    required super.version,
    required ModManifest manifest,
  }) : _manifestJson = jsonEncode(manifest.toJson()) {
    if (!_idEquals(id, manifest.id) || version != manifest.version) {
      throw const FormatException(
        'Selected identity must agree with its manifest.',
      );
    }
  }
  final String _manifestJson;
  ModManifest get manifest =>
      ModManifest.fromJson(jsonDecode(_manifestJson) as Map<String, Object?>);
}

final class InstallFacts {
  const InstallFacts({
    this.platform = '',
    this.architecture = '',
    this.contentTarget = '',
    this.gameVersion = '',
  });
  final String platform;
  final String architecture;
  final String contentTarget;
  final String gameVersion;
}

/// Exact selected installed packages; duplicate selections remain diagnosable.
final class EffectiveProfile {
  EffectiveProfile({
    required Iterable<ResolvedPackage> packages,
    required String profileId,
    required int revision,
    this.install = const InstallFacts(),
    Iterable<ResolvedPackage> disabledPackages = const [],
  }) : profileId = _token(profileId),
       revision = _integer(revision),
       packages = _boundedCopy(packages),
       disabledPackages = _boundedCopy(disabledPackages);
  final String profileId;
  final int revision;
  final List<ResolvedPackage> packages;
  final List<ResolvedPackage> disabledPackages;
  final InstallFacts install;
}
